README.md
# VGS BGM Decoder
- decodes: `VGS's .BGM file` to `PCM`
- platform free

## License
[The BSD 2-Clause License](https://github.com/suzukiplan/vgs-bgm-decoder/blob/master/LICENSE.txt)

## How to use
- Create Context
- Load BGM data (from file or memory)
  - _Get Meta Information (optional)_
  - Decode
  - _Get/Set Register (optional)_
- Release Context

## Create Context
#### prototyping
```
void* vgsdec_create_context();
```

#### return value
- `!NULL` : context
- `NULL` : error

## Load BGM data
#### prototyping
```
int vgsdec_load_bgm_from_file(void* context, const char* path);
int vgsdec_load_bgm_from_memory(void* context, void* data, size_t size);
```

#### arguments
- `context` : context
- `path` : path of the .BGM or .vgs file
- `data` : pointer of memory area that BGM or .vgs file data stored.
- `size` : size of the .BGM or .vgs file

#### return value
- `0` : success
- `-1` : failed

## Get Meta Information
#### prototyping
```
struct VgsMetaHeader* vgsdec_get_meta_header(void* context);
struct VgsMetaData* vgsdec_get_meta_data(void* context, int index);
```

#### arguments
- `context` : context
- `index` : index of the meta data

#### return value
- NULL: if meta header or data is not exist.
- !NULL: meta header or data

## Decode (BASIC)
#### prototyping
```
void vgsdec_execute(void* context, void* buffer, size_t size);
```

#### arguments
- `context` : context
- `buffer` : stores decoded PCM data
- `size` : size of `buffer`

#### remarks
- this function stores decoded PCM data to the buffer.
- The format of the PCM are fixed by the following:
  - sampling rate: 22050Hz
  - bit rate: 16bit
  - channels: 1 (monoral)
- ex: 1 second = 44100 byte _(22050 * 16 * 1 / 8)_
- use to sequentially call

## Decode (ASYNC)
_Async decoder function provides only for the senior programmer that fully understanding both of multiple threading mechanism and `vgsdec.c`. Please DO NOT USE, if you cannot the both or the either._

#### prototyping
```
int vgsdec_async_start(void* context);
int vgsdec_async_enqueue(void* context, void* buffer, size_t size, void (*callback)(void* context, void* buffer, size_t size));
void vgsdec_async_stop(void* context);
```

- `vgsdec_async_start` function
  - start async decoder thread
- `vgsdec_async_enqueue` function
  - enqueue a decode task
- `callback` function:
  - called by async decoder thread after decoded.
- `vgsdec_async_stop` function
  - stop async decoder thread

#### arguments
- `context` : context
- `buffer` : input/output buffer
- `size` : size of `buffer`
- `callback` : function pointer of callback after decoded

#### return value
- `0` : success
- `-1` : failed

#### remarks
- you must not call `vgsdec_execute` while running async decoder thread.

## Get/Set Register
#### prototyping
```
int vgsdec_get_value(void* context, int type);
void vgsdec_set_value(void* context, int type, int value);
```

#### arguments
- `context` : context
- `type` : register type
- `value` : register value

#### return value
- register value

#### type
|define|get|set|description|
|:---|:---:|:---:|:---|
|`VGSDEC_REG_KEY_{0~5}`|o|-|get the [scale value](#scale-value) of Ch{0~5}|
|`VGSDEC_REG_TONE_{0~5}`|o|-|get the [tone value](#tone-value) of Ch{0~5}|
|`VGSDEC_REG_VOL_{0~5}`|o|-|get the average volume of Ch{0~5}|
|`VGSDEC_REG_PLAYING`|o|-|current playing status: `0` = not playing, `1` = playing|
|`VGSDEC_REG_INDEX`|o|-|get the playing note-index|
|`VGSDEC_REG_LOOP_INDEX`|o|-|get the note index of the loop-label (acyclic song: -1)|
|`VGSDEC_REG_LENGTH`|o|-|get the count of the notes|
|`VGSDEC_REG_TIME`|o|o|get or set the playing time _(1sec = 22050)_|
|`VGSDEC_REG_LOOP_TIME`|o|-|get the time of the loop-label _(1sec = 22050)_|
|`VGSDEC_REG_TIME_LENGTH`|o|-|get the time-length _(1sec = 22050)_|
|`VGSDEC_REG_LOOP_COUNT`|o|-|get the loop-count|
|`VGSDEC_REG_RESET`|-|o|set none-zero: reset to the state after loading the data.|
|`VGSDEC_REG_FADEOUT`|-|o|set none-zero: start fadeout.|
|`VGSDEC_REG_FADEOUT_COUNTER`|o|-|get [fadeout counter](#fadeout-counter)|
|`VGSDEC_REG_VOLUME_RATE_{0~5}`|o|o|get or set the channel volume rate (0~100)|
|`VGSDEC_REG_VOLUME_RATE`|o|o|get or set the master volume rate (0~100)|
|`VGSDEC_REG_SYNTHESIS_BUFFER`|o|o|set none-zero: does not clear the buffer when called `vgsdec_execute`|
|`VGSDEC_REG_KEYON_{0~5}`|o|-|get key-on status of Ch{0~5}: `0` = key-off, `1` = key-on|
|`VGSDEC_REG_MUTE_{0~5}`|o|o|get or set the mute of Ch{0~5}: `0` = sound, `1` = mute|
|`VGSDEC_REG_ADD_KEY_{0~5}`|o|o|get or set the [scale up/down](#scale-updown) of Ch{0~5}|
|`VGSDEC_REG_KOBUSHI`|o|o|get or set the Kobushi mode: `0` = not kobushi, `1` = kobushi|

#### scale value
|Octave|`A`|`A#`|`B`|`C`|`C#`|`D`|`D#`|`E`|`F`|`F#`|`G`|`G#`|
|:---|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|:---:|
|0|0|1|2|3|4|5|6|7|8|9|10|11|
|1|12|13|14|15|16|17|18|19|20|21|22|23|
|2|24|25|26|27|28|29|30|31|32|33|34|35|
|3|36|37|38|39|40|41|42|43|44|45|46|47|
|4|48|49|50|51|52|53|54|55|56|57|58|59|
|5|60|61|62|63|64|65|66|67|68|69|70|71|
|6|72|73|74|75|76|77|78|79|80|81|82|83|
|7|84|-|-|-|-|-|-|-|-|-|-|-|

#### scale up/down
- set VGSDEC_REG_ADD_KEY = 5: C major -> F major
- set VGSDEC_REG_ADD_KEY = -1: C major -> B major
- please exclude rhythm channel _(but specific rhythm channel is impossible...)_

#### tone value
|value|name|
|---|---|
|0|Sin wave|
|1|Square ware|
|2|Saw wave|
|3|Noize wave|

#### fadeout counter
##### means of counter value
- `0` : noop
- `1` ~ `99` : in-progress
- `100` : finished

##### mechanism
- set `VGSDEC_REG_FADEOUT` --> `VGSDEC_REG_FADEOUT_COUNT` to be `1`
- `VGSDEC_REG_FADEOUT_COUNT` increments per __1024Hz__ _(about 0.04643990929705 seconds)_ , if not `0` or `100`
- `VGSDEC_REG_FADEOUT_COUNT` equals 100 --> `VGSDEC_REG_PLAYING` to be zero

## Release Context
#### prototyping
```
void vgsdec_release_context(void* context);
```

#### arguments
- `context` : context

## Build
### example decoder
```
$ cd example
$ make
```

### DLL (Windows only)
```
# nmake /f makedll.mak
```


CHANGES.md
# Changes

## Version 1.09
- bugfix: sometime crush if using KOBUSHI mode

## Version 1.08
- corrected warning in the 64bit platforms

## Version 1.07
- bugfix: crush when setting KOBUSHI mode

## Version 1.06
- added getter of the KOBUSHI mode
- bugfix: reset playing time when context_reset
- bugfix: reset time length before loading a song from memory

## Version 1.05
- bugfix: reset tone when context_reset
- added KOBUSHI mode

how to use:
```
vgsdec_set_value(context, VGSDEC_REG_KOBUSHI, 1);
```

_Kobushi occurs when the pitch of the singer's voice fluctuates irregularly within one scale degree: This compares with vibrato, which vibrates in a regular cycle._

## Version 1.04
- splits the declaration and code in miniz: https://github.com/suzukiplan/vgs-bgm-decoder/issues/31

## Version 1.03
- need include guard in vgsdec_internal.h: https://github.com/suzukiplan/vgs-bgm-decoder/issues/29

## Version 1.02
- correct the notation shaking: data -> buffer: https://github.com/suzukiplan/vgs-bgm-decoder/issues/27

## Version 1.01
- do not declare `#define __stdcall` in the vgsdec.h: https://github.com/suzukiplan/vgs-bgm-decoder/issues/25

## Version 1.00
- support async decoding: https://github.com/suzukiplan/vgs-bgm-decoder/issues/2
- performance optimizing: https://github.com/suzukiplan/vgs-bgm-decoder/issues/17 
- complete getter/setter: https://github.com/suzukiplan/vgs-bgm-decoder/issues/19
- added test: https://github.com/suzukiplan/vgs-bgm-decoder/issues/20

## Version 0.03 (beta)
- support windows: https://github.com/suzukiplan/vgs-bgm-decoder/issues/15 

## Version 0.02 (beta)
- support change volume rate: https://github.com/suzukiplan/vgs-bgm-decoder/issues/3
- support synthesis mode: https://github.com/suzukiplan/vgs-bgm-decoder/issues/7
- support loading .vgs file: https://github.com/suzukiplan/vgs-bgm-decoder/issues/8
- bugfix: https://github.com/suzukiplan/vgs-bgm-decoder/issues/4
- bugfix: https://github.com/suzukiplan/vgs-bgm-decoder/issues/11

## Version 0.01 (beta)
- first beta version
- there is a possibility of a destructive change



test/test-mt.c
#include <pthread.h>
#include <stdio.h>
#include <string.h>
#include "vgsdec.h"

static void* decoder(void* args);

int main(int argc, char* argv[])
{
    pthread_t tid[10];
    int i;

    if (argc < 2) {
        return 1;
    }

    for (i = 0; i < 10; i++) {
        if (pthread_create(&tid[i], NULL, decoder, argv[1])) {
            puts("pthread_create error");
            return 255;
        }
    }

    for (i = 0; i < 10; i++) {
        pthread_join(tid[i], NULL);
    }
    puts("ALL-DONE");
    return 0;
}

static void* decoder(void* args)
{
    const char* path = (const char*)args;
    char buf[1024];
    void* context;

    /* initialize context */
    context = vgsdec_create_context();
    if (NULL == context) {
        puts("vgsdec_create_context error.");
        return NULL;
    }

    /* load bgm data */
    if (vgsdec_load_bgm_from_file(context, path)) {
        puts("load error.");
        vgsdec_release_context(context);
        return NULL;
    }

    /* decoding loop */
    while (vgsdec_get_value(context, VGSDEC_REG_PLAYING) && vgsdec_get_value(context, VGSDEC_REG_LOOP_COUNT) == 0) {
        vgsdec_execute(context, buf, sizeof(buf));
    }

    /* waiting for the end of fadeout if looped */
    if (vgsdec_get_value(context, VGSDEC_REG_LOOP_COUNT)) {
        vgsdec_set_value(context, VGSDEC_REG_FADEOUT, 1);
        while (vgsdec_get_value(context, VGSDEC_REG_PLAYING)) {
            vgsdec_execute(context, buf, sizeof(buf));
        }
    }
    vgsdec_release_context(context);
    puts("DONE");
    return NULL;
}


test/test-reload.c
#include <stdio.h>
#include <string.h>
#include "vgsdec.h"

int main(int argc, char* argv[])
{
    char buf[1024];
    void* context;
    int i, j;

    if (argc < 2) {
        return 1;
    }

    /* initialize context */
    for (j = 0; j < 2; j++) {
        context = vgsdec_create_context();
        if (NULL == context) {
            puts("vgsdec_create_context error.");
            return 255;
        }

        for (i = 0; i < 2; i++) {
            /* load bgm data */
            if (vgsdec_load_bgm_from_file(context, argv[1])) {
                puts("load error.");
                vgsdec_release_context(context);
                return 2;
            }

            /* decoding loop */
            while (vgsdec_get_value(context, VGSDEC_REG_PLAYING) && vgsdec_get_value(context, VGSDEC_REG_LOOP_COUNT) == 0) {
                vgsdec_execute(context, buf, sizeof(buf));
            }

            /* waiting for the end of fadeout if looped */
            if (vgsdec_get_value(context, VGSDEC_REG_LOOP_COUNT)) {
                vgsdec_set_value(context, VGSDEC_REG_FADEOUT, 1);
                while (vgsdec_get_value(context, VGSDEC_REG_PLAYING)) {
                    vgsdec_execute(context, buf, sizeof(buf));
                }
            }
        }
        vgsdec_release_context(context);
    }
    puts("DONE");
    return 0;
}


test/test-general.c
#include <stdio.h>
#include <string.h>
#include "vgsdec.h"

int main(int argc, char* argv[])
{
    char buf[1024];
    void* context;

    if (argc < 2) {
        return 1;
    }

    /* initialize context */
    context = vgsdec_create_context();
    if (NULL == context) {
        puts("vgsdec_create_context error.");
        return 255;
    }

    /* load bgm data */
    if (vgsdec_load_bgm_from_file(context, argv[1])) {
        puts("load error.");
        vgsdec_release_context(context);
        return 2;
    }

    /* decoding loop */
    while (vgsdec_get_value(context, VGSDEC_REG_PLAYING) && vgsdec_get_value(context, VGSDEC_REG_LOOP_COUNT) == 0) {
        vgsdec_execute(context, buf, sizeof(buf));
    }

    /* waiting for the end of fadeout if looped */
    if (vgsdec_get_value(context, VGSDEC_REG_LOOP_COUNT)) {
        vgsdec_set_value(context, VGSDEC_REG_FADEOUT, 1);
        while (vgsdec_get_value(context, VGSDEC_REG_PLAYING)) {
            vgsdec_execute(context, buf, sizeof(buf));
        }
    }
    vgsdec_release_context(context);
    puts("DONE");
    return 0;
}


src/vgsdec.h
/* (C)2016, SUZUKI PLAN.
 *----------------------------------------------------------------------------
 * Description: VGSDEC - header
 *    Platform: Common
 *      Author: Yoji Suzuki (SUZUKI PLAN)
 *----------------------------------------------------------------------------
 */
#ifndef INCLUDE_VGSDEC_H
#define INCLUDE_VGSDEC_H

#define VGSDEC_REG_KEY_0 0x10
#define VGSDEC_REG_KEY_1 0x11
#define VGSDEC_REG_KEY_2 0x12
#define VGSDEC_REG_KEY_3 0x13
#define VGSDEC_REG_KEY_4 0x14
#define VGSDEC_REG_KEY_5 0x15
#define VGSDEC_REG_TONE_0 0x20
#define VGSDEC_REG_TONE_1 0x21
#define VGSDEC_REG_TONE_2 0x22
#define VGSDEC_REG_TONE_3 0x23
#define VGSDEC_REG_TONE_4 0x24
#define VGSDEC_REG_TONE_5 0x25
#define VGSDEC_REG_VOL_0 0x30
#define VGSDEC_REG_VOL_1 0x31
#define VGSDEC_REG_VOL_2 0x32
#define VGSDEC_REG_VOL_3 0x33
#define VGSDEC_REG_VOL_4 0x34
#define VGSDEC_REG_VOL_5 0x35
#define VGSDEC_REG_PLAYING 0x40
#define VGSDEC_REG_INDEX 0x50
#define VGSDEC_REG_LOOP_INDEX 0x51
#define VGSDEC_REG_LENGTH 0x52
#define VGSDEC_REG_TIME 0x60
#define VGSDEC_REG_LOOP_TIME 0x61
#define VGSDEC_REG_TIME_LENGTH 0x62
#define VGSDEC_REG_LOOP_COUNT 0x70
#define VGSDEC_REG_RESET 0x80
#define VGSDEC_REG_FADEOUT 0x90
#define VGSDEC_REG_FADEOUT_COUNTER 0xA0
#define VGSDEC_REG_VOLUME_RATE_0 0x100
#define VGSDEC_REG_VOLUME_RATE_1 0x101
#define VGSDEC_REG_VOLUME_RATE_2 0x102
#define VGSDEC_REG_VOLUME_RATE_3 0x103
#define VGSDEC_REG_VOLUME_RATE_4 0x104
#define VGSDEC_REG_VOLUME_RATE_5 0x105
#define VGSDEC_REG_VOLUME_RATE 0x200
#define VGSDEC_REG_SYNTHESIS_BUFFER 0x300
#define VGSDEC_REG_KEYON_0 0x400
#define VGSDEC_REG_KEYON_1 0x401
#define VGSDEC_REG_KEYON_2 0x402
#define VGSDEC_REG_KEYON_3 0x403
#define VGSDEC_REG_KEYON_4 0x404
#define VGSDEC_REG_KEYON_5 0x405
#define VGSDEC_REG_MUTE_0 0x500
#define VGSDEC_REG_MUTE_1 0x501
#define VGSDEC_REG_MUTE_2 0x502
#define VGSDEC_REG_MUTE_3 0x503
#define VGSDEC_REG_MUTE_4 0x504
#define VGSDEC_REG_MUTE_5 0x505
#define VGSDEC_REG_ADD_KEY_0 0x600
#define VGSDEC_REG_ADD_KEY_1 0x601
#define VGSDEC_REG_ADD_KEY_2 0x602
#define VGSDEC_REG_ADD_KEY_3 0x603
#define VGSDEC_REG_ADD_KEY_4 0x604
#define VGSDEC_REG_ADD_KEY_5 0x605

#ifdef __cplusplus
extern "C" {
#endif

struct VgsMetaHeader {
    char eyecatch[8];
    char format[8];
    char genre[32];
    unsigned char num;
    unsigned char reserved1[3];
    unsigned short secIi;
    unsigned short secIf;
    unsigned short secLi;
    unsigned short secLf;
    unsigned int reserved2;
};

struct VgsMetaData {
    unsigned short year;
    unsigned short aid;
    unsigned short track;
    unsigned short reserved;
    char album[56];
    char song[64];
    char team[32];
    char creator[32];
    char right[32];
    char code[32];
};

#ifdef _WIN32
void* __stdcall vgsdec_create_context();
int __stdcall vgsdec_load_bgm_from_file(void* context, const char* path);
int __stdcall vgsdec_load_bgm_from_memory(void* context, void* data, size_t size);
void __stdcall vgsdec_execute(void* context, void* buffer, size_t size);
int __stdcall vgsdec_get_value(void* context, int type);
void __stdcall vgsdec_set_value(void* context, int type, int value);
void __stdcall vgsdec_release_context(void* context);
struct VgsMetaHeader* __stdcall vgsdec_get_meta_header(void* context);
struct VgsMetaData* __stdcall vgsdec_get_meta_data(void* context, int index);
int __stdcall vgsdec_async_start(void* context);
int __stdcall vgsdec_async_enqueue(void* context, void* buffer, size_t size, void (*callback)(void* context, void* buffer, size_t size));
void __stdcall vgsdec_async_stop(void* context);
#else
void* vgsdec_create_context();
int vgsdec_load_bgm_from_file(void* context, const char* path);
int vgsdec_load_bgm_from_memory(void* context, void* data, size_t size);
void vgsdec_execute(void* context, void* buffer, size_t size);
int vgsdec_get_value(void* context, int type);
void vgsdec_set_value(void* context, int type, int value);
void vgsdec_release_context(void* context);
struct VgsMetaHeader* vgsdec_get_meta_header(void* context);
struct VgsMetaData* vgsdec_get_meta_data(void* context, int index);
int vgsdec_async_start(void* context);
int vgsdec_async_enqueue(void* context, void* buffer, size_t size, void (*callback)(void* context, void* buffer, size_t size));
void vgsdec_async_stop(void* context);
#endif

#ifdef __cplusplus
};
#endif

#endif


src/vgsint.h
/* (C)2016, SUZUKI PLAN.
 *----------------------------------------------------------------------------
 * Description: VGSDEC - internal header
 *    Platform: Common
 *      Author: Yoji Suzuki (SUZUKI PLAN)
 *----------------------------------------------------------------------------
 */
#include <stdio.h>
#include <stdlib.h>
#include <stdarg.h>
#include <string.h>
#include <ctype.h>
#include <time.h>
#include <math.h>
#include "vgsdec.h"

#define SAMPLE_RATE 22050
#define SAMPLE_BITS 16
#define SAMPLE_CH 1
#define SAMPLE_BUFS 4410

#define NTYPE_ENV1 1
#define NTYPE_ENV2 2
#define NTYPE_VOL 3
#define NTYPE_MVOL 4
#define NTYPE_KEYON 5
#define NTYPE_KEYOFF 6
#define NTYPE_WAIT 7
#define NTYPE_PDOWN 8
#define NTYPE_JUMP 9
#define NTYPE_LABEL 10

struct _NOTE {
    unsigned char type;
    unsigned char op1;
    unsigned char op2;
    unsigned char op3;
    unsigned int val;
};

struct _PSGCH {
    short* tone;
    unsigned char vol;
    unsigned char key;
    unsigned char keyOn;
    unsigned char mute;
    unsigned int cur;
    unsigned int count;
    unsigned int env1;
    unsigned int env2;
    unsigned char toneT;
    unsigned char toneK;
    unsigned int pdown;
    unsigned int pcnt;
    int volumeRate;
};

struct _PSG {
    struct _NOTE* notes;
    unsigned char play;
    unsigned char mask;
    unsigned short mvol;
    unsigned int waitTime;
    short wav[6];
    int nidx;
    int stopped;
    unsigned int fade;
    unsigned int fcnt;
    struct _PSGCH ch[6];
    int mute;
    int loop;
    int fade2;
    unsigned int timeI;
    unsigned int timeL;
    unsigned int timeP;
    int addKey[6];
    int addOff[6];
    int idxnum;
    int volumeRate;
};

extern short* TONE1[85];
extern short* TONE2[85];
extern short* TONE3[85];
extern short* TONE4[85];


src/vgsdec_internal.h
/* (C)2016, SUZUKI PLAN.
 *----------------------------------------------------------------------------
 * Description: VGS BGM Decoder - internal header
 *    Platform: Common
 *      Author: Yoji Suzuki (SUZUKI PLAN)
 *----------------------------------------------------------------------------
 */
#ifndef INCLUDE_VGSDEC_INTERNAL_H
#define INCLUDE_VGSDEC_INTERNAL_H
#ifdef _WIN32
#include <winsock2.h>
#include <windows.h>
#include <process.h>
#define inline
#else
#include <pthread.h>
#include <unistd.h>
#include <limits.h>
#include <arpa/inet.h>
#endif
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "vgsdec.h"
#define MINIZ_HEADER_FILE_ONLY
#include "miniz.h"

#define NTYPE_ENV1 1
#define NTYPE_ENV2 2
#define NTYPE_VOL 3
#define NTYPE_MVOL 4
#define NTYPE_KEYON 5
#define NTYPE_KEYOFF 6
#define NTYPE_WAIT 7
#define NTYPE_PDOWN 8
#define NTYPE_JUMP 9
#define NTYPE_LABEL 10

#define MAX_NOTES 65536

struct _NOTE {
    unsigned char type;
    unsigned char op1;
    unsigned char op2;
    unsigned char op3;
    unsigned int val;
};

struct _PSGCH {
    short* toneS;
    int* tone;
    int toneA;
    int toneR;
    int* toneM;
    int toneMA;
    int toneMR;
    int toneMP;
    unsigned char vol;
    unsigned char keyOn;
    unsigned char mute;
    unsigned int cur;
    unsigned int count;
    unsigned int env1;
    unsigned int env2;
    unsigned char toneT;
    unsigned char toneK;
    unsigned int pdown;
    unsigned int pcnt;
    int volumeRate;
};

struct _VGS_QDATA {
    struct _VGS_QDATA* next;
    void (*callback)(void* context, void* buffer, size_t size);
    void* context;
    void* buffer;
    size_t size;
};

struct _VGS_QUEUE {
    volatile int enabled;
    volatile int count;
#ifdef _WIN32
    CRITICAL_SECTION cs;
    uintptr_t tid;
#else
    pthread_mutex_t mt;
    pthread_attr_t ta;
    pthread_t tid;
#endif
    struct _VGS_QDATA* head;
    struct _VGS_QDATA* tail;
};

struct _VGSCTX {
#ifdef _WIN32
    CRITICAL_SECTION cs;
#else
    pthread_mutex_t mt;
#endif
    struct _NOTE notes[MAX_NOTES];
    struct VgsMetaHeader* mhead;
    struct VgsMetaData** mdata;
    unsigned char play;
    unsigned char mask;
    unsigned short mvol;
    unsigned int waitTime;
    unsigned int hz;
    int wav[6];
    int nidx;
    int stopped;
    unsigned int fade;
    unsigned int fcnt;
    struct _PSGCH ch[6];
    int mute;
    int loop;
    int fade2;
    unsigned int timeI;
    unsigned int timeL;
    unsigned int timeP;
    int addKey[6];
    int addOff[6];
    int loopIdx;
    int idxnum;
    int volumeRate;
    int synthesis;
    struct _VGS_QUEUE queue;
};

extern short* TONE1[85];
extern short* TONE2[85];
extern short* TONE3[85];
extern short* TONE4[85];

static void reset_context(struct _VGSCTX* c);
static void lock_context(struct _VGSCTX* c);
static void unlock_context(struct _VGSCTX* c);
static void lock_queue(struct _VGSCTX* c);
static void unlock_queue(struct _VGSCTX* c);
static inline void set_note(struct _VGSCTX* c, unsigned char cn, unsigned char t, unsigned char n);
static inline int get_next_note(struct _VGSCTX* c);
static void jump_time(struct _VGSCTX* c, int sec);
static size_t extract_meta_data(struct _VGSCTX* c, void* data, size_t size);
static void release_meta_data(struct _VGSCTX* c);
static void msleep(int ms);
#ifdef _WIN32
static unsigned __stdcall async_manager(void* context);
#else
static void* async_manager(void* context);
#endif

#endif


example/decoder.c
/* basic decoder sample */
#ifdef _WIN32
#include <io.h>
#else
#include <unistd.h>
#define _open open
#define _write write
#define _lseek lseek
#define _close close
#endif
#include <fcntl.h>
#include <stdio.h>
#include <string.h>
#include "vgsdec.h"

#ifndef O_BINARY
#define O_BINARY 0
#endif

struct DataHeader {
    char riff[4];
    unsigned int fsize;
    char wave[4];
    char fmt[4];
    unsigned int bnum;
    unsigned short fid;
    unsigned short ch;
    unsigned int sample;
    unsigned int bps;
    unsigned short bsize;
    unsigned short bits;
    char data[4];
    unsigned int dsize;
};

static char* sec(int hz)
{
    static char result[8];
    int s, m;
    s = hz / 22050;
    m = s / 60;
    s %= 60;
    sprintf(result, "%02d:%02d", m, s);
    return result;
}

static void dump(void* context);

int main(int argc, char* argv[])
{
    char buf[1024];
    void* context;
    FILE* wav;
    struct DataHeader dh;

    if (argc < 3) {
        puts("usage: decoder bgm_file wav_file");
        return 1;
    }

    /* initialize context */
    context = vgsdec_create_context();
    if (NULL == context) {
        puts("vgsdec_create_context error.");
        return 255;
    }

    /* load bgm data */
    if (vgsdec_load_bgm_from_file(context, argv[1])) {
        puts("load error.");
        vgsdec_release_context(context);
        return 2;
    }

    /* open wave file */
    wav = fopen(argv[2], "wb");
    if (NULL == wav) {
        puts("open error.");
        vgsdec_release_context(context);
        return 3;
    }

    /* initialize wave header */
    memset(&dh, 0, sizeof(dh));
    strncpy(dh.riff, "RIFF", 4);
    strncpy(dh.wave, "WAVE", 4);
    strncpy(dh.fmt, "fmt ", 4);
    strncpy(dh.data, "data", 4);
    dh.bnum = 16;
    dh.fid = 1;
    dh.ch = 1;
    dh.sample = 22050;
    dh.bps = 44100;
    dh.bsize = 2;
    dh.bits = 16;
    dh.dsize = 0;
    fwrite(&dh, 1, sizeof(dh), wav);

    dump(context);

    /* decoding loop */
    while (vgsdec_get_value(context, VGSDEC_REG_PLAYING) && vgsdec_get_value(context, VGSDEC_REG_LOOP_COUNT) == 0) {
        vgsdec_execute(context, buf, sizeof(buf));
        fwrite(buf, sizeof(buf), 1, wav);
        dh.dsize += sizeof(buf);
    }

    /* waiting for the end of fadeout if looped */
    if (vgsdec_get_value(context, VGSDEC_REG_LOOP_COUNT)) {
        vgsdec_set_value(context, VGSDEC_REG_FADEOUT, 1);
        while (vgsdec_get_value(context, VGSDEC_REG_PLAYING)) {
            vgsdec_execute(context, buf, sizeof(buf));
            fwrite(buf, sizeof(buf), 1, wav);
            dh.dsize += sizeof(buf);
        }
    }

    fclose(wav);
    vgsdec_release_context(context);

    /* update wave header */
    dh.fsize = dh.dsize + sizeof(dh) - 8;
    {
        int fd = _open(argv[2], O_RDWR | O_BINARY);
        if (-1 != fd) {
            _lseek(fd, 0, 0);
            _write(fd, &dh, sizeof(dh));
            _close(fd);
        }
    }
    return 0;
}

static void dump(void* context)
{
    int i, v;
    struct VgsMetaHeader* mhead;
    struct VgsMetaData* mdata;

    /* meta header if exist */
    mhead = vgsdec_get_meta_header(context);
    printf("META-HEADER: ");
    if (NULL != mhead) {
        printf("\n");
        printf(" - format: %s\n", mhead->format);
        printf(" - genre: %s\n", mhead->genre);
        printf(" - data count: %d\n", (int)mhead->num);
    } else {
        printf("n/a\n");
    }

    /* meta data if exist */
    for (i = 0; NULL != (mdata = vgsdec_get_meta_data(context, i)); i++) {
        printf("META-DATA #%d:\n", i + 1);
        printf(" - year: %d\n", (int)mdata->year);
        printf(" - aid: %d\n", (int)mdata->aid);
        printf(" - track: %d\n", (int)mdata->track);
        printf(" - album: %s\n", mdata->album);
        printf(" - song: %s\n", mdata->song);
        printf(" - team: %s\n", mdata->team);
        printf(" - creator: %s\n", mdata->creator);
        printf(" - right: %s\n", mdata->right);
        printf(" - code: %s\n", mdata->code);
    }

    /* length info */
    printf("NUMBER OF NOTES: %d\n", vgsdec_get_value(context, VGSDEC_REG_LENGTH));
    printf("LOOP-INDEX: %d\n", vgsdec_get_value(context, VGSDEC_REG_LOOP_INDEX));
    printf("TIME: %s\n", sec(vgsdec_get_value(context, VGSDEC_REG_TIME_LENGTH)));
    printf("LOOP-START: %s\n", sec(vgsdec_get_value(context, VGSDEC_REG_LOOP_TIME)));

    /* channel volume */
    printf("CHANNEL-VOLUME:");
    for (i = 0; i < 6; i++) {
        vgsdec_set_value(context, VGSDEC_REG_VOLUME_RATE_0 + i, 100);
        if (i) printf(",");
        printf(" CH%d=%d", i, vgsdec_get_value(context, VGSDEC_REG_VOLUME_RATE_0 + i));
    }
    printf("\n");

    /* master volume */
    vgsdec_set_value(context, VGSDEC_REG_VOLUME_RATE, 100);
    printf("MASTER-VOLUME: %d\n", vgsdec_get_value(context, VGSDEC_REG_VOLUME_RATE));

    /* channel mute */
    printf("CHANNEL-MUTE:");
    for (i = 0; i < 6; i++) {
        vgsdec_set_value(context, VGSDEC_REG_MUTE_0 + i, 0);
        /* vgsdec_set_value(context, VGSDEC_REG_MUTE_0 + i, (i == 0 || i == 1) ? 1 : 0); // ex: mute melody */
        if (i) printf(",");
        printf(" CH%d=%d", i, vgsdec_get_value(context, VGSDEC_REG_MUTE_0 + i));
    }
    printf("\n");

    /* scale up/down */
    printf("SCALE:");
    for (i = 0; i < 6; i++) {
        // vgsdec_set_value(context, VGSDEC_REG_ADD_KEY_0 + i, -1);
        if (i) printf(",");
        v = vgsdec_get_value(context, VGSDEC_REG_ADD_KEY_0 + i);
        printf(" CH%d=%s%d", i, v < 0 ? "" : "+", v);
    }
    printf("\n");

    /* set KOBUSHI mode */
    vgsdec_set_value(context, VGSDEC_REG_KOBUSHI, 1);
}
